janus_server/src/binary_utils/job_driver.rs

```diff
@@ -0,0 +1,414 @@
+                Err(err) => {
+                    error!(?err, "Couldn't acquire jobs");
+                    // Go ahead and step job discovery delay in this error case to ensure we don't
+                    // tightly loop running transactions that will fail without any delay.
+                    job_discovery_delay = self.step_job_discovery_delay(job_discovery_delay);
+                    continue;
+                }
+            };
+            if acquired_jobs.is_empty() {
+                debug!("No jobs available");
+                job_discovery_delay = self.step_job_discovery_delay(job_discovery_delay);
+                continue;
+            }
+            info!(acquired_job_count = acquired_jobs.len(), "Acquired jobs");
+
+            // Start up tasks for each acquired job.
+            job_discovery_delay = Duration::ZERO;
+            for (acquired_job, lease_expiry) in acquired_jobs {
+                task::spawn({
+                    // We acquire a semaphore in the job-discovery task rather than inside the new
+                    // job-stepper task to ensure that acquiring a permit does not race with
+                    // checking how many permits we have available in the next iteration of this
+                    // loop, to maintain the invariant that this task is the only place we acquire
+                    // permits.
+                    //
+                    // Unwrap safety: we have seen that at least `acquired_jobs.len()` permits are
+                    // available, and this task is the only task that acquires permits.
+                    let permit = Arc::clone(&sem).try_acquire_owned().unwrap();
+                    let this = Arc::clone(&self);
+                    let acquired_job = Arc::new(acquired_job);
+                    let acquired_job_clone = acquired_job.clone();
+
+                    async move {
+                        info!(?lease_expiry, "Stepping job");
+                        match time::timeout(
+                            this.effective_lease_duration(lease_expiry),
+                            (this.job_stepper)(
+                                this.datastore.clone(),
+                                acquired_job.clone(),
+                                this.job_stepper_context.clone(),
+                            ),
+                        )
+                        .await
+                        {
+                            Ok(Ok(_)) => {
+                                debug!("Job stepped")
+                            }
+
+                            Ok(Err(err)) => {
+                                error!(?err, "Couldn't step job")
+                            }
+
+                            Err(err) => error!(?err, "Stepping job timed out"),
+                        }
+                        drop(permit);
+                    }
+                    .instrument(info_span!("Job stepper", acquired_job = ?acquired_job_clone))
```

Comment:
This is minor, but we could avoid putting `acquired_job` in an Arc and cloning it if we constructed the span first, moved the acquired job into the future, then annotated the future with the span, and adjusted signatures to match.

Reply:
Huh, I had thought that `info_span!` would take ownership of fields passed into it, but I guess it must be invoking `std::fmt::Debug::fmt` on the field at the time the span gets created instead of holding on to the value. I implemented your suggestion in #188. Thanks!